[Catalyst] News/Books sometimes crash under `-[WKContentView _didHandleKeyEvent:eventWasHandled:]`

whsieh opened the pull request:
#### f6389c81ebed0cf721df03e2b0b9df977da3f502
<pre>
[Catalyst] News/Books sometimes crash under `-[WKContentView _didHandleKeyEvent:eventWasHandled:]`
<a href="https://bugs.webkit.org/show_bug.cgi?id=289413">https://bugs.webkit.org/show_bug.cgi?id=289413</a>
<a href="https://rdar.apple.com/130233195">rdar://130233195</a>

Reviewed by Abrar Rahman Protyasha.

This is a speculative fix for the crashes in <a href="https://rdar.apple.com/130233195">rdar://130233195</a>; the existing code invokes the web key
event completion handler inside of the `removeFirstMatching` lambda, but if the completion handler
ends up calling into code that mutates `_keyWebEventHandlers`, the vector can end up in a bad state.

Adjust the code to be robust against this, by removing the matching `KeyEventAndCompletionBlock`
from `_keyWebEventHandlers` *before* calling into the completion block.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _didHandleKeyEvent:eventWasHandled:]):

Canonical link: <a href="https://commits.webkit.org/291856@main">https://commits.webkit.org/291856@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7057e15b08f0b843ecbe9afb7a45f1538a3de47b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94206 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13793 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3552 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99217 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44733 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14093 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22223 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71864 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29202 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97208 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10455 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85058 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52208 "Found 1 new API test failure: /WPE/TestWebKitWebView:/webkit/WebKitWebView/save (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10145 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44051 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80377 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2833 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101262 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21258 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15475 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80870 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21510 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81071 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80252 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20010 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24795 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2162 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14445 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21242 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26421 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20929 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24389 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22670 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -7672,12 +7672,21 @@ - (void)_didHandleKeyEvent:(::WebEvent *)event eventWasHandled:(BOOL)eventWasHan
-            return false;
-        entry.completionBlock(event, eventWasHandled);
-        return true;
+    auto indexOfHandlerToCall = _keyWebEventHandlers.findIf([&](auto& entry) {
```
Nit: just capture event?

whsieh replied:
Will do — thanks for the review!

whsieh committed:
[Catalyst] News/Books sometimes crash under `-[WKContentView _didHandleKeyEvent:eventWasHandled:]`
https://bugs.webkit.org/show_bug.cgi?id=289413
rdar://130233195

Reviewed by Abrar Rahman Protyasha.

This is a speculative fix for the crashes in rdar://130233195; the existing code invokes the web key
event completion handler inside of the `removeFirstMatching` lambda, but if the completion handler
ends up calling into code that mutates `_keyWebEventHandlers`, the vector can end up in a bad state.

Adjust the code to be robust against this, by removing the matching `KeyEventAndCompletionBlock`
from `_keyWebEventHandlers` *before* calling into the completion block.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _didHandleKeyEvent:eventWasHandled:]):

Canonical link: https://commits.webkit.org/291856@main
